### PR TITLE
SQL: Make parsing of date more lenient

### LINKED ISSUE #49379: SQL: implicit CASTing between DATEs and DATETIMEs
We have a hard CASTing rule for DATEs and DATETIMEs, ie `CAST(CAST(date AS STRING) AS DATE)` will fail with `SqlIllegalArgumentException[cannot cast [2004-06-04T00:00:00.000Z] to [date]: Text '2004-06-04T00:00:00.000Z' could not be parsed, unparsed text found at index 10]` because a date and a time is not actually a date.
Should we loosen this hard rule for DATEs and DATETIMEs? Should we do implicit CASTing?

### PR DESCRIPTION
**Pull request by matriv**
Make the parsing of date more lenient
 - as an escaped literal: `{d '2020-02-10[[T| ]10:20[:30][.123456789][tz]]'}`
 - cast a string to a date: `CAST(2020-02-10[[T| ]10:20[:30][.123456789][tz]]' AS DATE)`

Closes: #49379

### CODE REVIEW

**Comment by matriv** [comment]
I'm not sure about the escaped literal enhancement, since MS-SQL docs only mention the `yyyy-mm-dd` format: https://docs.microsoft.com/en-us/sql/connect/jdbc/using-sql-escape-sequences?view=sql-server-ver15#date-and-time-literals

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/type/SqlDataTypeConverterTests.java [hunk 1]
```diff
@@ -172,13 +172,19 @@ public void testConversionToDate() {
...
-            assertEquals(date(-1672531200000L), conversion.convert("1917-01-01"));
-            assertEquals(date(18000000L), conversion.convert("1970-01-01"));
+            assertEquals(date(1581292800000L), conversion.convert("2020-02-10T10:20"));
+            assertEquals(date(-125908819200000L), conversion.convert("-2020-02-10T10:20:30.123"));
+            assertEquals(date(1581292800000L), conversion.convert("2020-02-10T10:20:30.123456789"));
 
-            // double check back and forth conversion
+            assertEquals(date(1581292800000L), conversion.convert("2020-02-10 10:20"));
+            assertEquals(date(-125908819200000L), conversion.convert("-2020-02-10 10:20:30.123"));
+            assertEquals(date(1581292800000L), conversion.convert("2020-02-10 10:20:30.123456789"));
```
Comment (bpintea) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/type/SqlDataTypeConverterTests.java]: For sake of completeness, I'd suggest a test without the duration separator, but with a timezone (or a `Z`).

**Review by bpintea — APPROVED** [review]
LGTM

**Review by astefan — APPROVED** [review]
LGTM. Left some further small comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/DateUtils.java [hunk 2]
```diff
@@ -91,7 +115,17 @@ public static ZonedDateTime asDateTime(long millis, ZoneId id) {
...
      */
     public static ZonedDateTime asDateOnly(String dateFormat) {
-        return LocalDate.parse(dateFormat, ISO_LOCAL_DATE).atStartOfDay(UTC);
+        int separatorIdx = dateFormat.indexOf('-');
+        if (separatorIdx == 0) { // negative year
+            separatorIdx = dateFormat.indexOf('-', separatorIdx + 1);
+        }
+        separatorIdx = dateFormat.indexOf('-', separatorIdx + 1) + 3;
+        // Avoid index out of bounds - it will lead to DateTimeParseException anyways
+        if (separatorIdx >= dateFormat.length() || dateFormat.charAt(separatorIdx) == 'T') {
```
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/DateUtils.java]: 2 lines above we do `+3` to move directly after the day part in the string.
If the string passed is something like: `2020-12-` then we have moved passed the string's length and `dateFormat.charAt()` will throw OutOfBounds exception.

I forgot to add a test for that, will do.

**Pull request merged**